File: backend/src/PaymentLab.Api/Domain/Transaction.cs
namespace PaymentLab.Api.Domain;

public sealed class Transaction
{
    public required string Id { get; init; }

    public required string MerchantId { get; init; }

    /// <summary>Principal charged to the cardholder.</summary>
    public required Money Gross { get; init; }

    /// <summary>Merchant discount rate fee retained by the PJP at settlement.</summary>
    public required Money MerchantFee { get; init; }

    public required DateTimeOffset AuthorizedAt { get; init; }

    /// <summary>
    /// Already masked at ingestion (6-4 format). The full PAN never enters this service.
    /// RED ZONE: must never be logged, echoed in errors, or widened.
    /// </summary>
    public required string MaskedPan { get; init; }

    public TransactionStatus Status { get; set; } = TransactionStatus.Settled;

    public List<Refund> Refunds { get; init; } = new();

    public Money TotalRefunded()
    {
        var total = Money.ZeroLike(Gross);
        foreach (var refund in Refunds)
        {
            total = total.Add(refund.Amount);
        }

        return total;
    }

    public Money TotalFeeReversed()
    {
        var total = Money.ZeroLike(MerchantFee);
        foreach (var refund in Refunds)
        {
            total = total.Add(refund.FeeReversed);
        }

        return total;
    }

    /// <summary>Principal still available to refund.</summary>
    public Money RefundableRemaining() => Gross.Subtract(TotalRefunded());
}

File: backend-node/src/domain/transaction.ts
import { Money } from './money.js';

/** Port of Domain/TransactionStatus.cs. */
export type TransactionStatus = 'Settled' | 'PartiallyRefunded' | 'Refunded' | 'Failed';

/** Port of Domain/Refund.cs. */
export interface Refund {
  id: string;
  transactionId: string;
  amount: Money;
  feeReversed: Money;
  reason: string;
  idempotencyKey: string;
  createdAt: Date;
}

/** Port of Domain/Transaction.cs. */
export class Transaction {
  readonly id: string;
  readonly merchantId: string;
  readonly gross: Money;
  readonly merchantFee: Money;
  readonly maskedPan: string;
  readonly createdAt: Date;
  status: TransactionStatus;
  readonly refunds: Refund[] = [];

  constructor(params: {
    id: string;
    merchantId: string;
    gross: Money;
    merchantFee: Money;
    maskedPan: string;
    createdAt: Date;
    status: TransactionStatus;
  }) {
    this.id = params.id;
    this.merchantId = params.merchantId;
    this.gross = params.gross;
    this.merchantFee = params.merchantFee;
    this.maskedPan = params.maskedPan;
    this.createdAt = params.createdAt;
    this.status = params.status;
  }

  totalRefunded(): Money {
    return this.refunds.reduce((sum, r) => sum.add(r.amount), Money.zeroLike(this.gross));
  }

  totalFeeReversed(): Money {
    return this.refunds.reduce((sum, r) => sum.add(r.feeReversed), Money.zeroLike(this.merchantFee));
  }

  /** Remaining principal that can still be refunded. Compare against THIS, never against `gross` directly. */
  refundableRemaining(): Money {
    return this.gross.subtract(this.totalRefunded());
  }
}
